domain/application/service/application.go

```diff
@@ -1288,3 +1355,135 @@ func encodeApplicationConfig(cfg config.ConfigAttributes, charmConfig charm.Conf
 	}
 	return encodedConfig, nil
 }
+
+func validateSecretConfig(chCfg internalcharm.Config, cfg internalcharm.Settings) error {
+	for name, value := range cfg {
+		option, ok := chCfg.Options[name]
+		if !ok {
+			// This should never happen.
+			return errors.NotValidf("unknown option %q", name)
+		}
+		if option.Type == "secret" {
+			uriStr, ok := value.(string)
+			if !ok {
+				return applicationerrors.InvalidSecretConfig
+			}
+			if uriStr == "" {
+				return nil
+			}
+			_, err := secrets.ParseURI(uriStr)
+			return errors.Annotatef(err, "invalid secret URI for option %q", name)
+		}
+	}
+	return nil
+}
+
+func decodeCharmOrigin(origin application.CharmOrigin) (corecharm.Origin, error) {
+	decodedSource, err := decodeCharmSource(origin.Source)
+	if err != nil {
+		return corecharm.Origin{}, internalerrors.Errorf("decoding charm source: %w", err)
+	}
+
+	decodePlatform, err := decodePlatform(origin.Platform)
+	if err != nil {
+		return corecharm.Origin{}, internalerrors.Errorf("decoding platform: %w", err)
+	}
+
+	decodedChannel, err := decodeChannel(origin.Channel)
+	if err != nil {
+		return corecharm.Origin{}, internalerrors.Errorf("decoding channel: %w", err)
+	}
+
+	return corecharm.Origin{
+		Source:   decodedSource,
+		Channel:  decodedChannel,
+		Platform: decodePlatform,
+	}, nil
+}
+
+func decodeCharmSource(source charm.CharmSource) (corecharm.Source, error) {
+	switch source {
+	case charm.CharmHubSource:
+		return corecharm.CharmHub, nil
+	case charm.LocalSource:
+		return corecharm.Local, nil
+	default:
+		return "", internalerrors.Errorf("unknown charm source type %q", source)
```

Comment:
tiny nit: in https://github.com/juju/juju/blob/main/domain/application/state/state.go#L1405 we return a 
```
		return "", fmt.Errorf("unsupported charm source: %d", source)
```
Maybe we should do the same here or update state as well?